Increment wallet key manager index during recovery

philipr-za opened the pull request:
## Description
Currently, when recovery is performed the key manager index is not incremented as outputs are found. This means that when new transactions are performed the key manager produces previously used spending keys for the new outputs, reusing private keys. We also did not apply the correct script private key to the recovered output. This wasn’t a problem because we are using `Nop` scripts but for any script that requires this key to be correct we would have had a problem.

This PR adds the logic that when an output is rewound during recovery that we find the key manager index for the recovered spending key, update the key manager index to one more than the highest index found during recovery and also use this index to apply the correct script private key to the recovered output.

This PR also contains a but of neatness refactoring in the Output Manager Service.

## How Has This Been Tested?
Existing tests

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [x] I have squashed my commits into a single commit.

delta1 reviewed (changes requested):
Looks great! The update to `package.json` means that you'll have to commit `package-lock.json` as part of this change, `npm ci` installs directly from the lockfile which is out of date now

philipr-za commented:
> Looks great! The update to `package.json` means that you'll have to commit `package-lock.json` as part of this change, `npm ci` installs directly from the lockfile which is out of date now

Done

delta1 reviewed (approved):
LGTM

SWvheerden commented on base_layer/wallet/src/output_manager_service/master_key_manager.rs:
```diff
@@ -0,0 +1,222 @@
+    pub async fn get_seed_words(&self) -> Result<Vec<String>, OutputManagerError> {
+        Ok(from_secret_key(
+            self.utxo_key_manager.lock().await.master_key(),
+            &MnemonicLanguage::English,
```
nit: Should this not rather be a config option for `MasterKeyManager`. 
This seems like a smell, hard-coded option within a function. 
I would rather say let `MasterKeyManager` have a field language and hard code it in there for now.

philipr-za replied:
Fair enough, gave this function a argument to specify the language and added it to the Output Manager Config and its a default now.

SWvheerden commented on base_layer/wallet/src/output_manager_service/recovery/standard_outputs_recoverer.rs:
```diff
@@ -0,0 +1,153 @@
+
+    /// Attempt to rewind all of the given transaction outputs into unblinded outputs. If they can be rewound then add
+    /// them to the database and increment the key manager index
+    pub async fn scan_and_recover_outputs(
```
```suggestion
    pub async fn scan_recover_and_add_outputs(
```

This function now scans, recovers, and adds the outputs it finds. 
I am not completely for this as this function now does more than one thing, but we should at least update the name to reflect that.

philipr-za replied:
I will argue that the verb `recover` is the process of adding it to the database. The method does to two things, `scans` which is going through and trying to rewind an output and `recover` which is adding it to the db and updating the keymanager. I thought about splitting the second part out but decided that I couldn't think of a use case for every allowing these to happen in isolation and that the interface should only allow them to be done together.

delta1 replied:
I agree with @philipr-za , "recover" here already means "recover for spending" which means add them to your wallet.

SWvheerden reviewed (commented):
Looks good, 2 nits thou.

delta1 commented on base_layer/wallet/src/output_manager_service/handle.rs:
```diff
@@ -103,7 +103,7 @@ impl fmt::Display for OutputManagerRequest {
             GetPublicRewindKeys => write!(f, "GetPublicRewindKeys"),
             FeeEstimate(_) => write!(f, "FeeEstimate"),
-            RewindOutputs(_, _) => write!(f, "RewindAndImportOutputs"),
+            ScanForRecoverableOutputs(_, _) => write!(f, "ScanAndImportRecoverableOutputs"),
```
just noticed these don't match, not sure it's an issue though

philipr-za replied:
fixed

philipr-za committed:
Increment wallet key manager index during recovery

Currently, when recovery is performed the key manager index is not incremented as outputs are found. This means that when new transactions are performed the key manager produces previously used spending keys for the new outputs, reusing private keys. We also did not apply the correct script private key to the recovered output. This wasn’t a problem because we are using `Nop` scripts but for any script that requires this key to be correct we would have had a problem.

This PR adds the logic that when an output is rewound during recovery that we find the key manager index for the recovered spending key, update the key manager index to one more than the highest index found during recovery and also use this index to apply the correct script private key to the recovered output.

This PR also contains a but of neatness refactoring in the Output Manager Service.
